File: packages/yakr-core/src/yakr_core/hybrid_pq.py
from __future__ import annotations

from pqcrypto.kem.ml_kem_768 import (
    CIPHERTEXT_SIZE,
    PLAINTEXT_SIZE,
    PUBLIC_KEY_SIZE,
    SECRET_KEY_SIZE,
    decrypt,
    encrypt,
    generate_keypair,
)

from yakr_core.crypto import hkdf_derive

HYBRID_MASTER_INFO = b"yakr/v0.6/hybrid-master"
HYBRID_PQ_CAPABILITY = "hybrid_pq"
PQ_REKEY_MAX_MESSAGES = 10_000
PQ_REKEY_MAX_AGE_MS = 7 * 24 * 60 * 60 * 1000


def kem_generate_keypair() -> tuple[bytes, bytes]:
    public_key, secret_key = generate_keypair()
    return public_key, secret_key


def kem_encapsulate(public_key: bytes) -> tuple[bytes, bytes]:
    if len(public_key) != PUBLIC_KEY_SIZE:
        raise ValueError("invalid ML-KEM-768 public key length")
    ciphertext, shared_secret = encrypt(public_key)
    return ciphertext, shared_secret


def kem_decapsulate(secret_key: bytes, ciphertext: bytes) -> bytes:
    if len(secret_key) != SECRET_KEY_SIZE:
        raise ValueError("invalid ML-KEM-768 secret key length")
    if len(ciphertext) != CIPHERTEXT_SIZE:
        raise ValueError("invalid ML-KEM-768 ciphertext length")
    return decrypt(secret_key, ciphertext)


def derive_hybrid_master(
    *,
    identity_shared: bytes,
    ephemeral_shared: bytes,
    pq_secret: bytes,
    transcript_hash: bytes,
) -> bytes:
    if len(pq_secret) != PLAINTEXT_SIZE:
        raise ValueError("invalid PQ shared secret length")
    x_secret = identity_shared + ephemeral_shared
    return hkdf_derive(x_secret + pq_secret, HYBRID_MASTER_INFO, salt=transcript_hash)


def needs_pq_rekey(
    *,
    hybrid: bool,
    session_started_at_ms: int,
    messages_sent: int,
    now_ms: int | None = None,
) -> bool:
    if not hybrid:
        return False
    import time

    now = int(time.time() * 1000) if now_ms is None else now_ms
    age_exceeded = now - session_started_at_ms >= PQ_REKEY_MAX_AGE_MS
    count_exceeded = messages_sent >= PQ_REKEY_MAX_MESSAGES
    return age_exceeded or count_exceeded

File: rust/yakr-core/src/hybrid_pq.rs
//! ML-KEM-768 hybrid post-quantum helpers.

#![allow(deprecated)]

use ml_kem::{
    array::Array,
    kem::{Decapsulate, Encapsulate, Generate, KeyExport, Kem, TryKeyInit},
    Ciphertext, DecapsulationKey, EncapsulationKey, ExpandedKeyEncoding, MlKem768,
};

pub const HYBRID_PQ_CAPABILITY: &str = "hybrid_pq";
pub const PQ_REKEY_MAX_MESSAGES: u64 = 10_000;
pub const PQ_REKEY_MAX_AGE_MS: u64 = 7 * 24 * 60 * 60 * 1000;

// ML-KEM-768 sizes (FIPS 203 / Python `pqcrypto.kem.ml_kem_768`).
pub const KEM_PUBLIC_KEY_SIZE: usize = 1184;
pub const KEM_SECRET_KEY_SIZE: usize = 2400;
pub const KEM_CIPHERTEXT_SIZE: usize = 1088;
pub const KEM_SHARED_SECRET_SIZE: usize = 32;

pub fn kem_generate_keypair() -> (Vec<u8>, Vec<u8>) {
    let (dk, ek) = MlKem768::generate_keypair();
    (
        ek.to_bytes().as_slice().to_vec(),
        dk.to_expanded_bytes().as_slice().to_vec(),
    )
}

pub fn kem_encapsulate(public_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>), String> {
    if public_key.len() != KEM_PUBLIC_KEY_SIZE {
        return Err("invalid ML-KEM-768 public key length".into());
    }
    let key = Array::<u8, _>::clone_from_slice(public_key);
    let ek = EncapsulationKey::<MlKem768>::new(&key).map_err(|_| "invalid public key")?;
    let (ct, ss) = ek.encapsulate();
    Ok((ct.as_slice().to_vec(), ss.as_slice().to_vec()))
}

pub fn kem_decapsulate(secret_key: &[u8], ciphertext: &[u8]) -> Result<Vec<u8>, String> {
    if secret_key.len() != KEM_SECRET_KEY_SIZE {
        return Err("invalid ML-KEM-768 secret key length".into());
    }
    if ciphertext.len() != KEM_CIPHERTEXT_SIZE {
        return Err("invalid ML-KEM-768 ciphertext length".into());
    }
    let expanded = Array::<u8, _>::clone_from_slice(secret_key);
    let dk = DecapsulationKey::<MlKem768>::from_expanded_bytes(&expanded)
        .map_err(|_| "invalid secret key")?;
    let ss = dk
        .decapsulate_slice(ciphertext)
        .map_err(|_| "invalid ciphertext")?;
    Ok(ss.as_slice().to_vec())
}

pub fn needs_pq_rekey(hybrid: bool, session_started_at_ms: u64, messages_sent: u64) -> bool {
    if !hybrid {
        return false;
    }
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64;
    let age_exceeded = now.saturating_sub(session_started_at_ms) >= PQ_REKEY_MAX_AGE_MS;
    let count_exceeded = messages_sent >= PQ_REKEY_MAX_MESSAGES;
    age_exceeded || count_exceeded
}
